### PR TITLE
Cleanup type headers in preparation for supporting all types.

### PR DESCRIPTION
**Pull request by aentinger**

### CODE REVIEW

File: src/types/uavcan/_register/Access_1_0.h [hunk 1]
```diff
@@ -0,0 +1,542 @@
+// This is an AUTO-GENERATED UAVCAN DSDL data type implementation. Curious? See https://uavcan.org.
+// You shouldn't attempt to edit this file.
+//
+// Checking this file under version control is not recommended unless it is used as part of a high-SIL
+// safety-critical codebase. The typical usage scenario is to generate it as part of the build process.
```
Comment (pavel-kirienko) [inline src/types/uavcan/_register/Access_1_0.h:5]: Are you sure this is a sensible move?

> A question one is likely to ask here is: Why don't you ship precompiled regulated DSDL together with the tool? Indeed, that would be trivial to do, but we avoid that on purpose to emphasize our commitment to supporting vendor-specific and regulated DSDL at the same level. In the past we used to give regulated namespaces special treatment, which caused our users to acquire misconceptions about the purpose of DSDL. Specifically, there have been forks of the standard namespace extended with vendor-specific types, which is harmful to the ecosystem.

Surely there should be a way to use vendor-specific types. If that is the case, then probably the same principles could be applied to the standard data types as well?

Comment (aentinger) [reply src/types/uavcan/_register/Access_1_0.h:5]: >Are you sure this is a sensible move?

No? I get the feeling that I'm misunderstanding the UAVCAN concept a little. So the regulated types in `uavcan` are not supposed to be used individually and should only be used by incorporating them into a DSDL composite type representing some sensor/actuator with the workflow being: 1) create DSDL for your sensor/actuator using uavcan regulated types, 2) generate C headers for your sensor/actuator and 3) use those C headers in your application? If so I'm a bit at loss how to "Arduino-fy" this workflow.

Comment (pavel-kirienko) [reply src/types/uavcan/_register/Access_1_0.h:5]: Not quite, no. It is possible to use the standard types by themselves, but I would recommend ensuring that vendor- or application-specific types are as easy to use as the standard ones, otherwise it may foster misconceptions as I wrote in the excerpt above. I think a better approach would be to somehow enable Arduino users to easily generate headers for arbitrary namespaces, whether it's `uavcan`, `reg`, or `whatever`. Shall that turn out to be impossible, I am willing to consider having a trivial website wrapping Nunavut that lets you upload a DSDL namespace (or a link to GitHub repository containing one) and spits generated code back.

I would like to tag @thirtytwobits because he might have relevant ideas.

Comment (aentinger) [reply src/types/uavcan/_register/Access_1_0.h:5]: Providing a web-service around the DSDL generator might be a sensible move. You've correctly picked up on my concern that requiring the typical Arduino user to self-generate the C header files might be one step too much. Also it's contradictory with allowing the usage of the library out of the box without any further setup steps.

Furthermore there's the issue of some "convenience" [typedefs](https://github.com/107-systems/107-Arduino-UAVCAN/blob/all-uavcan/src/types/uavcan/node/Heartbeat.1.0.hpp#L30-L36) and [functions](https://github.com/107-systems/107-Arduino-UAVCAN/blob/all-uavcan/src/types/uavcan/node/Heartbeat.1.0.hpp#L76-L79) which I fail to see how to code-generate them. Therefore my idea was to periodically generate the C headers and merge changes into this repository, preserving the convenience functionality and Arduino wrappers.

But really, I'm open for anything that might lead to a self-contained and easy-to-use Arduino library (wrapper of UAVCAN).

Comment (pavel-kirienko) [reply src/types/uavcan/_register/Access_1_0.h:5]: > Furthermore there's the issue of some "convenience" typedefs and functions which I fail to see how to code-generate them.

Excellent examples. The first one is to be eventually [supported by DSDL language itself](https://github.com/UAVCAN/specification/issues/52), but for now, we have to limit ourselves to regular constants. The other one is an interesting shortcut that some might describe as an abuse of operator overloading, but at any rate, it is trivial to automate using Nunavut template language. DSDL in its current form is a pretty spartan language that has some potential for improvement.

Regarding the web service -- can we, perchance, recruit some resources from your team? From our side, we can dispatch our glorious sysadmin Clyde to get the web service functional and deployed but he will need help with the business logic.

Comment (aentinger) [reply src/types/uavcan/_register/Access_1_0.h:5]: I believe a first step would be to augment nunavut to generate the required Arduino headers such as [Heartbeat.1.0.hpp](https://github.com/107-systems/107-Arduino-UAVCAN/blob/all-uavcan/src/types/uavcan/node/Heartbeat.1.0.hpp) and [ExecuteCommand.1.0.hpp](https://github.com/107-systems/107-Arduino-UAVCAN/blob/all-uavcan/src/types/uavcan/node/ExecuteCommand.1.0.hpp). This is something that I could help with, despite many personal misgivings on `Python`. With regards to the web service I can't really promise anything as no one within [107-systems](https://github.com/107-systems) comes with real web expertise. Perhaps we can discuss a bit more on the next weekly dev call.

Comment (generationmake) [reply src/types/uavcan/_register/Access_1_0.h:5]: I think the typical Arduino user wants fast and easy solutions. He won't modify anything in the libraries or generate some code and include it in his project. He just wants to copy some code from the internet into his sketch and see it running.
I think we should support this. With this pull request I think this is possible. The tons of headers are not nice but at the moment I don't see any other way.
Additionally we shoud give the more professional users a way to create their own namespaces. As an Arduino project grows bigger and better at some point it should be able to switch to a own namespace and a DSDL for the project. Maybe we can support this by showing in examples and documentation how to do this.

Comment (pavel-kirienko) [reply src/types/uavcan/_register/Access_1_0.h:5]: > I think the typical Arduino user wants fast and easy solutions.

This is true, but I don't think the proposed way of supplying pre-generated headers for a subset of data types is going to enable fast and easy solutions. For comparison, this is as if you were using a programming language that allows you to use classes from its standard library but does not allow you to define your own. Yes, you don't need anything else to print "hello world", but any non-trivial application is likely to demand more from the language.

Doing it this way may incentivize bad practices and create a skewed perception of what UAVCAN is. I recommend that we enable this volunteer who showed up just this morning to help us with the web service idea and then see whether it is easy to integrate it into the typical Arduino workflow: https://forum.uavcan.org/t/an-easy-to-use-web-service-for-compiling-dsdl/1074/2?u=pavel.kirienko

Comment (aentinger) [reply src/types/uavcan/_register/Access_1_0.h:5]: I do believe both of your inputs are valid and merit satisfying them in some way. Most users of this library probably won't ever stray from the types already defined so those would be served by pre-generated and periodically updated header files. The other (and probably more advanced) users would want to generate their own headers for their own types. Also I __get__ it that Pavel want's to retain the cleanliness of UAVCAN and not repeat situations that came up in V0.

How about the following proposal: We move the already existing types directly into the examples making use of them. They will still be checked into this repository but in the example sections so new users can immediately fire up the Arduino IDE and **do something**. For more advanced user we'll document how to use the yet to be designed web service (congratulations in finding a web-dev in such short time span, wish I could find a Python coder too) in the projects README. Furthermore the generated types for GNSS sensor and ToF sensor will in fact be checking in with all other code, otherwise we it wouldn't work out of the box. What do you say about that?

Comment (pavel-kirienko) [reply src/types/uavcan/_register/Access_1_0.h:5]: > We move the already existing types directly into the examples making use of them. They will still be checked into this repository but in the example sections so new users can immediately fire up the Arduino IDE and do something. For more advanced user we'll document how to use the yet to be designed web service in the projects README. 

Okay, let's do it. But can you please make it very explicit in the docs (so that even an inattentive reader would notice) that there's more to DSDL than the regulated namespaces? We don't want to accidentally repeat the same mistakes here.

Comment (aentinger) [reply src/types/uavcan/_register/Access_1_0.h]: After given the matter some more thought I'm loath to actually copy the headers to the examples right now and duplicate code this way (e.g. we would need all the Heartbeat related files for both `UAVCAN-Heartbeat-Publish` and `UAVCAN-Heartbeat-Subscribe` - symbolic links are forbidden within Arduino libraries).

Otherwise I'd rather wait for C++ header generation by nunavut (@thirtytwobits do you have an estimation when those might be available?) and then integrate those C++ headers to work seamlessly with this library OR to have a extension of nunavut which generates those currently manually encoded Arduino wrappers (see https://github.com/UAVCAN/nunavut/issues/179). Unfortunately for the 2nd option we'd need to find someone who's willing and capable to do the nunavut code generator coding with no one currently in sight.

@pavel-kirienko @generationmake what are your thoughts here?

PS: In order to progress with [UAVCAN-GNSS-Node](https://github.com/107-systems/UAVCAN-GNSS-Node) it could be thinkable to directly check-in generated C headers for the GNSS node and manually implemented Arduino wrappers within that repository and leave them out of 107-Arduino-UAVCAN for now.

Comment (thirtytwobits) [reply src/types/uavcan/_register/Access_1_0.h]: Starting work on C++ generation here -> https://github.com/thirtytwobits/nunavut/tree/issue/91

Comment (thirtytwobits) [reply src/types/uavcan/_register/Access_1_0.h]: The Arduino-way to do this sort of thing is to be very device-specific and have a set of DSDL for a given device as a separate library that utilizes a core library (see Adafruit's core GFX library and the individual libraries for various displays they sell as an example). As such, I'd provide the pre-generated headers for the core standard in a central UAVCAN/Arduino library and create libraries for each UAVCAN device supported that had pre-generated headers for that device and examples that utilized them. 

Most Arduino projects are started with the example sketches that are distributed with a library so these device-specific libraries tend to be the entry-points that quietly pull-in the core libraries they extend.

This scheme also lets you push update notifications to the IDE which is better then having people generate code on their own since any bugs we find in the serialization logic can be fixed and the users notified by the library update mechanism.

In-short, I do think distributed pre-generated headers is appropriate for Arduino. It is a specific ecosystem where software development best practices don't always apply.

Comment (pavel-kirienko) [reply src/types/uavcan/_register/Access_1_0.h]: How would an Arduno user transpile custom DSDL in this case?

Comment (thirtytwobits) [reply src/types/uavcan/_register/Access_1_0.h]: Using your website I suppose but the most important thing about Arduino is the immediacy. You need to have a way to run a fully-functional example before you can expect users to move on to advance concepts like IDLs and transpilers.

Comment (aentinger) [reply src/types/uavcan/_register/Access_1_0.h]: Scott, thank you very much for providing your perspective on this. I'm agreeing with you that pre-generated headers are the way to go for Arduino (which is what I initially set out to do within this PR). Adopters of this library already have set out to create their [own wrappers](https://github.com/rotoye/107-Arduino-UAVCAN/blob/master/src/types/wrappers/SourceTs.0.1.hpp) so instead of having a rank growth of every adopter writing some wrappers and header to provide their types, why not providing them all from the source library (this one).

In short: I'm still pro storing C generated headers and manually (tiresome, but doable) Arduino wrappers within this library.
